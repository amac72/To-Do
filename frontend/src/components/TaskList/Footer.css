.footer {
    height: 150px;
    width: 100vw;
    background-image: none;
    background-repeat: repeat;
    background-attachment: scroll;
    background-position: 0% 0%;
    position: fixed;
    bottom: 0pt;
    pointer-events: none;
    white-space: pre;
}

.footer button {
    width: 17vmin;
    height: 17vmin;
    border-radius: 200px;
    box-shadow: 0 0 5px grey;
    font-size: 50px;
    pointer-events: all;
    opacity: 0.7;
    display: inline-block;
}

#reorder_tasks {
    font-size: 20px;
    position: fixed;
    bottom: 4vw;
    left: 4vw;
}

#add_task {
    padding-bottom: 14px;
    position: fixed;
    bottom: 4vw;
    right: 4vw;
}