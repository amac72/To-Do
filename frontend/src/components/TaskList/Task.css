.description {
    grid-area: task;
    border: none;
    font-size: 20px;
    width: auto;
}

.date {
    grid-area: date;
    border: none;
    font-size: 20px;
    color: grey;
    width: 150px;
}

.deleteButton {
    grid-area: delete;
    text-align: center;
    border-radius: 20px;
}

.completeButton {
    grid-area: check;
    text-align: center;
    border-radius: 20px;
    color: black;
    background-color: white;
    border: none;
    padding: none;
    box-shadow: none;
}

.completeButton:hover {
    background-color: white;
    color: black;
    border: none;
    box-shadow: none;
}

.completeButton:focus {
    background-color: white;
    color: black;
    border: none;
    box-shadow: none;
}

.completeButton:active {
    background-color: white;
    color: black;
    border: none;
    box-shadow: none !important;
}

.TaskFrame {
    display: grid;
    grid-template-areas:
        'check task task task task task delete'
        'check date date date date date delete';
    gap: 10px;
    padding: 7px;
    width: 75vmin;
    margin: auto;
    border: solid 1px gray;
    border-radius: 20px;
    box-shadow: 0 0 5px grey;
}