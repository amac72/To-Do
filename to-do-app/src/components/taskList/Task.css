.description {
    grid-area: task;
    border: none;
    font-size: 20px;
    width: auto;
}

.date {
    grid-area: date;
    border: none;
    font-size: 20px;
    color: grey;
    width: 150px;
}

.deleteButton {
    grid-area: delete;
    text-align: center;
    border-radius: 20px;
}

.TaskFrame {
    display: grid;
    grid-template-areas:
        'task task task task task delete'
        'date date date date date delete';
    gap: 10px;
    padding: 7px;
    width: 75vmin;
    margin: auto;
    border: solid 1px gray;
    border-radius: 20px;
    box-shadow: 0 0 5px grey;
}