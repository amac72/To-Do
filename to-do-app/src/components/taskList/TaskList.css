.bottom_of_page {
    height: 150px;
    width: 100%;
    background-image: none;
    background-repeat: repeat;
    background-attachment: scroll;
    background-position: 0% 0%;
    position: fixed;
    bottom: 0pt;
    left: 0pt;
    pointer-events: none;
}

.bottom_of_page button {
    width: 100px;
    height: 100px;
    margin: auto;
    border: solid 1px gray;
    border-radius: 200px;
    box-shadow: 0 0 5px grey;
    font-size: 50px;
    pointer-events: all;
}

#reorder_tasks {
    margin-left: 100px;
    margin-right: 120vmin;
    font-size: 20px;
}

#add_task {
    margin-right: 100px;
    padding-bottom: 10px;
}