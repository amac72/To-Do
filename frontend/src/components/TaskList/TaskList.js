import React, { useState } from 'react';
import { useLocation } from 'react-router-dom';
import Task from './Task';
import Footer from './Footer';
import axios from 'axios';

function TaskList(props) {
    const { taskList, setTaskList, url } = props;
    const [nextId, setNextId] = useState(taskList[taskList.length - 1].id + 1)

    function getStatus(location) {
        const pathname = location.pathname;
        let status = ""
        if (pathname === "/To-Do/completed") {
            status = true
        } else {
            status = false
        }
        return status
    }

    const status = getStatus(useLocation())

    function updateTask(value, fieldName, id) {
        const temp = [...taskList].map(task => {
            if (task.id === id) {
                axios.put(`${url}taskitems/${id}`, {
                    id: task.id,
                    description: fieldName === "description" ? value : task.description,
                    date: fieldName === "date" ? value : task.date,
                    completed: fieldName === "completed" ? value : task.completed,
                })
                    .then((response) => {
                        console.log(response.data);
                    })
                    .catch(error => console.error(`Error: ${error}`));
                return { ...task, [fieldName]: value }
            } else {
                return { ...task }
            }
        })
        setTaskList(temp)
    }

    function deleteTask(id) {
        if (window.confirm("Are you sure you want to delete this task?")) {
            axios.delete(`${url}taskitems/${id}`)
                .then((response) => {
                    console.log(response.data);
                })
                .catch(error => console.error(`Error: ${error}`));
            setTaskList(taskList.filter(item => item.id !== id));
        }
    }

    function addTask() {
        axios.post(`${url}taskitems`, {
            id: nextId,
            description: null,
            date: null,
            completed: false
        })
            .then((response) => {
                console.log(response.data);
            })
            .catch(error => console.error(`Error: ${error}`));

        const temp = [...taskList];
        temp.push(
            {
                id: nextId,
                description: "",
                date: null,
                completed: false
            }
        );
        setNextId(nextId + 1)
        setTaskList(temp);
    }

    function reorderTasks() {
        const temp = [...taskList]
        temp.sort((a, b) => {
            const dateA = new Date(a.date).getTime() || 0
            const dateB = new Date(b.date).getTime() || 0
            if (dateA === 0) {
                return 1
            } else if (dateB === 0) {
                return -1
            } else {
                return dateA - dateB
            }
        })
        setTaskList(temp);
    }

    return (
        <div>
            {taskList.filter(task => task.completed === status).map(task => {
                return (
                    <div key={task.id} data-testid={`task-${task.id}`}>
                        <Task task={task} updateTask={updateTask} deleteTask={() => deleteTask(task.id)} />
                        <br></br>
                    </div>
                )
            })}
            <Footer reorderTasks={() => reorderTasks()} addTask={() => addTask()} status={status} />
        </div>
    );
}

export default TaskList;