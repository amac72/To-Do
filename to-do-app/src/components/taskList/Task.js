import React from 'react';
import { Button } from 'react-bootstrap';
import './Task.css';

function Task(props) {
    function generateCheckbox(completed) {
        if (completed) {
            return (
                <Button className="completeButton" variant="secondary" onClick={() => props.updateTask('false', 'completed', props.task.id)}><svg xmlns="http://www.w3.org/2000/svg" width="20" height="20" fill="currentColor" className="bi bi-check2-square" viewBox="0 0 16 16">
                    <path d="M3 14.5A1.5 1.5 0 0 1 1.5 13V3A1.5 1.5 0 0 1 3 1.5h8a.5.5 0 0 1 0 1H3a.5.5 0 0 0-.5.5v10a.5.5 0 0 0 .5.5h10a.5.5 0 0 0 .5-.5V8a.5.5 0 0 1 1 0v5a1.5 1.5 0 0 1-1.5 1.5H3z" />
                    <path d="m8.354 10.354 7-7a.5.5 0 0 0-.708-.708L8 9.293 5.354 6.646a.5.5 0 1 0-.708.708l3 3a.5.5 0 0 0 .708 0z" />
                </svg></Button>
            )
        } else {
            return (
                <Button className="completeButton" variant="secondary" onClick={() => props.updateTask('true', 'completed', props.task.id)}><svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" className="bi bi-square" viewBox="0 0 16 16">
                    <path d="M14 1a1 1 0 0 1 1 1v12a1 1 0 0 1-1 1H2a1 1 0 0 1-1-1V2a1 1 0 0 1 1-1h12zM2 0a2 2 0 0 0-2 2v12a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V2a2 2 0 0 0-2-2H2z" />
                </svg></Button>
            )
        }
    }

    return (
        <div className="TaskFrame container-lg">
            <input className="description" onChange={(event) => props.updateTask(event.target.value, 'description', props.task.id)} value={props.task.description} placeholder="Enter task here" />
            <input type="date" className="date" onChange={(event) => props.updateTask(event.target.value, 'date', props.task.id)} value={props.task.date} />
            <Button className="deleteButton gray-button" variant="secondary" onClick={() => props.deleteTask()}><svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" className="bi bi-trash3-fill" viewBox="0 0 16 16">
                <path d="M11 1.5v1h3.5a.5.5 0 0 1 0 1h-.538l-.853 10.66A2 2 0 0 1 11.115 16h-6.23a2 2 0 0 1-1.994-1.84L2.038 3.5H1.5a.5.5 0 0 1 0-1H5v-1A1.5 1.5 0 0 1 6.5 0h3A1.5 1.5 0 0 1 11 1.5Zm-5 0v1h4v-1a.5.5 0 0 0-.5-.5h-3a.5.5 0 0 0-.5.5ZM4.5 5.029l.5 8.5a.5.5 0 1 0 .998-.06l-.5-8.5a.5.5 0 1 0-.998.06Zm6.53-.528a.5.5 0 0 0-.528.47l-.5 8.5a.5.5 0 0 0 .998.058l.5-8.5a.5.5 0 0 0-.47-.528ZM8 4.5a.5.5 0 0 0-.5.5v8.5a.5.5 0 0 0 1 0V5a.5.5 0 0 0-.5-.5Z" />
            </svg></Button>
            {generateCheckbox(props.task.completed)}
        </div>
    )
}

export default Task;