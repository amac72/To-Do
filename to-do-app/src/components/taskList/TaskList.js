import React, { useState } from 'react';
import Task from './Task';
import data from "./test.json";
import './TaskList.css';

function TaskList() {
    const [taskList, setTaskList] = useState(data);

    function deleteTask(id) {
        let i = 0
        if (window.confirm("Are you sure you want to delete this task?")) {
            for (i = 0; i < taskList.length; i++) {
                if (taskList[i].id === id) {
                    break;
                }
            }
            const temp = [...taskList];
            temp.splice(i, 1);
            setTaskList(temp);

            // setTaskList(taskList.filter(item => item.id !== id));
        }
    }

    function addTask() {
        const temp = [...taskList];
        let nextID = -1
        if (temp.length === 0) {
            nextID = 1
        } else {
            nextID = temp[temp.length - 1].id + 1
        }
        temp.push(
            {
                id: nextID,
                description: "",
                date: "",
            }
        );
        setTaskList(temp);
    }

    return (
        <div>
            {taskList.map(task => {
                return (
                    <div key={task.id}>
                        <Task task={task} onClick={() => deleteTask(task.id)} />
                        <br></br>
                    </div>
                )
            })}
            <div className="bottom_of_page"></div>
            <div className="footer">
                <div className="footer_contents">
                    <button className="add_task" onClick={addTask}>+</button>
                </div>
            </div>
        </div>
    );
}

export default TaskList;