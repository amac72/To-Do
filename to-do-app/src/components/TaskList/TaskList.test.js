import { render, screen, cleanup } from '@testing-library/react';
import renderer from "react-test-renderer";
import Task from "./Task";
import TaskList from './TaskList';

afterEach(() => {
    cleanup();
});

test('should render example task list', () => {
    const taskList = [
        {
            id: 1,
            description: "Clean apartment",
            date: "2022-04-27",
            completed: "no"
        }, {
            id: 2,
            description: "Pay rent",
            date: "2022-05-01",
            completed: "no"
        }
    ];
    render(<TaskList data={taskList} />)
    const taskListElement1 = screen.getByTestId('task-1');
    const taskListElement2 = screen.getByTestId('task-2');
    expect(taskListElement1).toBeInTheDocument();
    expect(taskListElement2).toBeInTheDocument();
})

test('should render footer even if list is empty', () => {
    const taskList = [];
    render(<TaskList data={taskList} />)
    const taskListElement = screen.getByTestId('footer-1');
    expect(taskListElement).toBeInTheDocument();
})

test('matches snapshot', () => {
    const taskList = [
        {
            id: 1,
            description: "Clean apartment",
            date: "2022-04-27",
            completed: "no"
        }, {
            id: 2,
            description: "Pay rent",
            date: "2022-05-01",
            completed: "no"
        }
    ];
    const tree = renderer.create(<TaskList data={taskList} />).toJSON();
    expect(tree).toMatchSnapshot();
})