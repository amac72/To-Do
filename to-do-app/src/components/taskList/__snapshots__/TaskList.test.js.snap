// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <div
    data-testid="task-1"
  >
    <div
      className="TaskFrame"
    >
      <input
        className="description"
        onChange={[Function]}
        placeholder="Enter task here"
        value="Clean apartment"
      />
      <input
        className="date"
        onChange={[Function]}
        type="date"
        value="2022-04-27"
      />
      <button
        className="deleteButton"
        onClick={[Function]}
      >
        Delete
      </button>
    </div>
    <br />
  </div>
  <div
    data-testid="task-2"
  >
    <div
      className="TaskFrame"
    >
      <input
        className="description"
        onChange={[Function]}
        placeholder="Enter task here"
        value="Pay rent"
      />
      <input
        className="date"
        onChange={[Function]}
        type="date"
        value="2022-05-01"
      />
      <button
        className="deleteButton"
        onClick={[Function]}
      >
        Delete
      </button>
    </div>
    <br />
  </div>
  <div
    className="bottom_of_page"
  />
  <div
    className="footer"
    data-testid="footer-1"
  >
    <div
      className="footer_contents"
    >
      <button
        className="add_task"
        onClick={[Function]}
      >
        +
      </button>
    </div>
  </div>
</div>
`;
